velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -378,57 +350,87 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
   aggregateRel->mutable_common()->mutable_direct();
 }
 
-void VeloxToSubstraitPlanConvertor::toSubstrait(
+void VeloxToSubstraitPlanConvertor::toSubstraitJoin(
     google::protobuf::Arena& arena,
-    const std::shared_ptr<const core::AbstractJoinNode>& joinNode,
-    ::substrait::JoinRel* joinRel) {
-  std::vector<core::PlanNodePtr> sources = joinNode->sources();
+    const std::shared_ptr<const core::HashJoinNode> joinNode,
+    ::substrait::Rel* rel) {
+  const auto& joinSources = joinNode->sources();
+  auto joinOutputRowType =
+      joinSources[0]->outputType()->unionWith(joinSources[1]->outputType());
+
+  if (joinNode->joinType() == core::JoinType::kLeftSemi) {
+    joinOutputRowType = joinSources[0]->outputType();
+  } else if (joinNode->joinType() == core::JoinType::kRightSemi) {
+    joinOutputRowType = joinSources[1]->outputType();
+  }
+  auto joinOutputTypeSize = joinOutputRowType->size();
+
+  // Insert a project rel for outputRowType of HashJoinNode.
+  auto projectRel = rel->mutable_project();
+  auto projectEmitRel = projectRel->mutable_common()->mutable_emit();
```

Comment:
what's this for?

Reply:
'outputRowType' in velox hashJoinNode can be part of field reference in either left or/and right table. so we have to insert a projectRel for this

Reply:
only used in Line 377? any other place I miss?